Friction options clean

EstherHeck opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [ ] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [ ] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

EstherHeck committed:
added friction rheology

EstherHeck committed:
modified repeat timestep to faster decrease dt-size

EstherHeck committed:
added compute_current_edot_ii in material-utilities

EstherHeck committed:
friction options are taken into account from visco_plastic

EstherHeck committed:
friction options are taken into account from visco_plastic

EstherHeck committed:
changed critical slip distance from double to function input

EstherHeck committed:
found wrong } from rebasing

EstherHeck committed:
RSF in plasticadditional outputs - needs some more work

EstherHeck committed:
deleted ifs that sent me errors

EstherHeck committed:
added more info to function error

EstherHeck committed:
consider min SR for radiation damping

EstherHeck committed:
regularized RSF only with positive a

EstherHeck committed:
added *2 in radiation damping, fixed arcsinh-friction, moved elastic shear moduli

EstherHeck committed:
started on new yield mechanism

EstherHeck committed:
included current_edot_ii in plastic additional outputs

EstherHeck committed:
comments and citations for new yield option tresca

EstherHeck committed:
outcommented assert for yield case with elasticity, ran make indent

EstherHeck committed:
assume always yielding in fault for tresca yield mechanism

EstherHeck committed:
use friction coefficient in tresca yield case

EstherHeck committed:
added prints to find composition error

EstherHeck committed:
index variables for fault and theta computed within parse_parameters

EstherHeck committed:
resolved fault and theta index problems

EstherHeck committed:
RDT within fault stength for tresca

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+       * Enumeration for selecting which type of friction dependence to use.
+       * For the type 'independent', the user-supplied internal angle of friction is used.
+       * For the type 'dynamic friction' the friction angle is rate dependent using
+       * Equation 13 from van Dinther et al. (2013).
```
Please provide the full references (i.e. journal, volume) here or in the documentation of the rheology model at the bottom of the .cc file.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+       * For the type 'independent', the user-supplied internal angle of friction is used.
+       * For the type 'dynamic friction' the friction angle is rate dependent using
+       * Equation 13 from van Dinther et al. (2013).
+       * For the type 'rate and state dependent friction'  the friction angle is calculated
```
```suggestion
       * For the type 'rate and state dependent friction', the friction angle is calculated
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+      /**
+       * Enumeration for selecting which type of friction dependence to use.
+       * For the type 'independent', the user-supplied internal angle of friction is used.
+       * For the type 'dynamic friction' the friction angle is rate dependent using
```
```suggestion
       * For the type 'dynamic friction', the friction angle is rate dependent following
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+        independent,
+        dynamic_friction,
+        rate_and_state_dependent_friction,
+        rate_and_state_dependent_friction_plus_linear_slip_weakening,
```
Please document the other Mechanisms as well

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+
+          /**
+           * A function that returns a ComponentMask, which indicates that the component
+           * associated with theta should be excluded during the volume fraction computation.
```
```suggestion
           * associated with the state variable theta should be excluded during the volume fraction computation.
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+           * associated with theta should be excluded during the volume fraction computation.
+           */
+          ComponentMask get_theta_composition_mask(ComponentMask composition_mask) const;
+
```
Please throughout the files format as 1 empty line between functions in the .h file, 3 empty lines between class functions in the .cc file

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+
+          /**
+           * A function that fills the reaction terms for the state variable theta in
+           * MaterialModelOutputs object that is handed over.
```
```suggestion
           * the MaterialModelOutputs object that is handed over.
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+                                            const double dte,
+                                            MaterialModel::MaterialModelOutputs<dim> &out) const;
+
+          /**for (unsigned int j=0; j < volume_fractions.size(); ++j)
```
rm this loop from the comment

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          /**
+           * A function that returns if the state variable theta is used.
+           */
+          bool get_use_theta() const;
```
I vaguely remember suggesting names for this earlier, sorry if I suggested this name. But if this function only returns whether theta is used or not, I would rename it to `use_theta`. Same for radiation damping. Then if-statements are easily understood, e.g. `if(use_theta()) ...`

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          bool get_use_radiation_damping() const;
+
+          /**
+           * Function to calculate depth-dependent a and b values for state dependent friction.
```
```suggestion
           * Function to calculate depth-dependent a and b values for state-dependent friction at a certain depth and for composition j.
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          std::pair<double,double>
+          calculate_depth_dependent_a_and_b(const Point<dim> &position, const int j) const;
+
+          /**
```
Update the alignment of the * above this and the following functions.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          calculate_depth_dependent_a_and_b(const Point<dim> &position, const int j) const;
+
+          /**
+          * Function that gets the critical slip distance at a certain position.
```
```suggestion
          * Function that gets the critical slip distance at a certain position for composition j. 
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          double get_critical_slip_distance(const Point<dim> &position, const int j) const;
+
+          /**
+          * A value for the effective normal stress on the fault that is used in Tresca friction
```
I don't think this is the Tresca criterion

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          */
+          double effective_normal_stress_on_fault;
+
+          unsigned int fault_composition_index;
```
Please provide a comment explaining both parameters, also making clear whether this is the field index or the volume_fraction index.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+
+          /**
+           * Parameter about what mechanism should be used for the friction dependence.
+           * Possible options: (rate-and-state) independent | dynamic friction | rate and state dependent
```
There are more options available, please update. Also, rm `(rate-and-state)` as that is not the name of the mechanism in the input file

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          Rheology::DruckerPragerParameters drucker_prager_parameters;
+
+          /**
+           * Parameter about what mechanism should be used for the friction dependence.
```
```suggestion
           * Select the mechanism to be used for the friction dependence.
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          std::vector<double> dynamic_angles_of_internal_friction;
+
+          /**
+           * The characteristic strain rate value, where the angle of friction takes the mean
```
```suggestion
           * The characteristic strain rate value at which the angle of friction is taken as the mean
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+           * The critical slip distance in rate and state friction. Used to calculate the state
+           * variable theta.
+           */
+          //double critical_slip_distance;
```
rm

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+
+          /**
+           * Arbitrary strain rate at which friction equals the reference friction angle in
+           * rate and state friction.
```
Can you add why it's called quasi static?

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          bool use_radiation_damping;
+
+          /**
+           * Parsed functions that specify a and b depth-dependence when using the Function
```
Is the Function method still something you can select? Or are functions always used? Please update the comment.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          std::unique_ptr<Functions::ParsedFunction<dim> > rate_and_state_parameter_b_function;
+
+          /**
+           * The coordinate representation to evaluate the function. Possible
```
```suggestion
           * The coordinate representation to evaluate the functions for a, b, and critical slip distance L. Possible
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+
+          /**
+           * The coordinate representation to evaluate the function. Possible
+           * choices are depth, cartesian and spherical.
```
Do you see a use case where it makes sense to have different coordinate systems to specify a, b, and L? Or can you simplify to one system for all three functions?

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+           */
+
+          /**
+           * Reference velocity for slip rate dependence of rate and state parameter a.
```
```suggestion
           * Reference velocity for slip-rate dependence of rate-and-state parameter a.
```

anne-glerum replied:
same for parameters below

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          Utilities::Coordinates::CoordinateSystem coordinate_system_L;
+
+          /**
+           * Slip rate dependent rate and state friction
```
```suggestion
           * Slip-rate dependent rate-and-state friction
```

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,272 @@
+          std::vector<double> effective_friction_factor;
+
+          /**
+           * The critical slip distance in rate and state friction. Used to calculate the state
```
```suggestion
           * The critical slip distance L in rate-and-state friction. Used to calculate the state
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+      FrictionOptions<dim>::
+      compute_dependent_friction_angle(const double current_edot_ii,
+                                       const unsigned int j,  // j is from a for-loop over volume fractions
+                                       const std::vector<double> &composition,
```
So `composition` are not the volume fractions but the original compositions? Please add that as well.

EstherHeck replied:
I hand over in.composition[i]. As I use it together with the volume_fractions index j, I guess in.composition also includes the background field and includes all original compositions not only volume_fractions. This is in analogy to the other parameters within calculate_isostrain_viscosities. But now that you asked this question, I am wondering in general why we loop over volume_fractions.size() and not over the non-masked compositions only, which should be faster. Is there a reason? It is done everywhere in visco-plastic as far as I remember. The masked fields like stress_xx or theta should just return a 0 anyway, as they are not part of the volume? So why do we loop over all fields?

anne-glerum replied:
`in.composition[i]` does not have the background field, but does indeed include all the 'raw' compositional field values. The index `j` is from the volume fractions, so it should not be directly used on `in.composition[i]`. You need `in.composition[i]` to get at theta, but as long as you use `theta_composition_index` I think it's fine. But do add a comment that `composition` is not volume_fractions but the compositional fields. 

I agree that instead of having volume_fractions = n_fields + 1, it could be n_fields + 1 - masked fields. This would require restructuring of the input logic (not specifying material properties for the masked fields) and the compute_volume_fractions function. The masked fields came way later than these structures, so I think it's just grown this way. It would break compatibility, but something to discuss!

anne-glerum reviewed (commented):
Hi Esther, I just went through the .h file, still need to do the .cc file. You can accept (if you want) the suggested changes online, but I've never tried it before.

EstherHeck committed:
added Lapusta time step files

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+              // friction vs strain rate curve is rather step-like or more gradual.
+              // mu  = mu_d + (mu_s - mu_d) / ( (1 + strain_rate_dev_inv2/dynamic_characteristic_strain_rate)^X );
+              // Angles of friction are used in radians within ASPECT. The coefficient of friction is the tangent of the internal angle of friction.
+              const double mu = effective_friction_factor[j] * std::tan(dynamic_angles_of_internal_friction[j])
```
Should the `effective_friction_factor` only apply to the first mu_d or to the whole mu you compute here? At the moment it's the first, this seems incorrect.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                                + (std::tan(current_friction) - std::tan(dynamic_angles_of_internal_friction[j]))
+                                / (1. + std::pow((current_edot_ii / dynamic_characteristic_strain_rate),
+                                                 dynamic_friction_smoothness_exponent));
+              current_friction = std::atan (mu);
```
We discussed this before, but did you check whether atan returns the phi in the correct quadrant? 
I'dd add an Assert (not AssertThrow) here that mu < 1 and 0 < current_friction < something.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+              // Cellsize is needed for theta and the friction angle
+              // For now, the used cells are non-deforming squares, so the edge length in the
+              // x-direction is representative of the cell size.
+              // TODO as the cell size is used to compute the slip velocity as cell_size * strain_rate,
```
Didn't Stephan say this was ok? Or maybe adjust it to say that in case of mesh deformation this length might not be representative

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                {
+                  cellsize = current_cell->extent_in_direction(0);
+                  // Calculate the state variable theta
+                  // theta_old loads theta from previous time step
```
```suggestion
                  // theta_old is taken from the current compositional field theta
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                  const double theta = compute_theta(theta_old, current_edot_ii, cellsize, critical_slip_distance);
+
+                  // Calculate effective friction according to Equation (4) in Sobolev and Muldashev (2017):
+                  // mu = mu_st + a ln(V/V_st) + b ln((theta Vst)/L)
```
```suggestion
                  // mu = mu_st + a ln(V/V_st) + b ln((theta V_st)/L)
```

anne-glerum replied:
Also, `V/V_st` suggests dividing a velocity by another velocity. Below you compute V as `current_edot_ii*cell_size`, which indeed gives a velocity. But `V_st` is called `quasi_static_strain_rate`, so suggests it's a strain rate. Why?

EstherHeck replied:
Thanks for spotting this! I would say, that this is not correct indeed! I guess I should also multiply the quasi_static_strain_rate with the cellsize everywhere?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+
+                  // Get the values for a and b
+                  double rate_and_state_parameter_a = calculate_depth_dependent_a_and_b(position,j).first;
+                  double rate_and_state_parameter_b = calculate_depth_dependent_a_and_b(position,j).second;
```
```suggestion
                  const double rate_and_state_parameter_b = calculate_depth_dependent_a_and_b(position,j).second;
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                  rate_and_state_parameter_a = rate_and_state_parameter_a + slope_s_for_a
+                                               * log10((ref_v_for_a + current_edot_ii * cellsize)/ref_v_for_a);
+
+                  const double dependent_critical_slip_distance = critical_slip_distance + slope_s_for_L
```
Why introduce a new parameter for L, but modify a, to compute the slip dependency?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+            {
+              break;
+            }
+            case slip_rate_dependent_rate_and_state_dependent_friction:
```
This function has a lot of duplicated code from rate_and_state_dependent_friction, can you avoid that somehow?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                  // Equation (7) from Sobolev and Muldashev (2017)
+                  const double theta = compute_theta(theta_old, current_edot_ii, cellsize, critical_slip_distance);
+
+                  // as RSF parameter a is used to divide current_friction will be nan without this if
```
```suggestion
                  // As we divide by RSF parameter a here, current_friction will be nan when a is zero
```

anne-glerum replied:
Also, what should happen when a=0? According to line 207, mu would be zero. Do you need to include this case?

EstherHeck replied:
yes, I need this case, because else all friction angles outside the fault will be nan. As currently I limit rate-and-state behaviour to the fault by setting a and b zero for all other materials. Which results in not modifying the friction angle for the classical log-formulation of RSF. I included that in the comment now.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                      // so conversion with tan/atan().
+                      current_friction = atan(effective_friction_factor[j]
+                                              * (rate_and_state_parameter_a
+                                                 * asinh((current_edot_ii * cellsize ) / (2 * quasi_static_strain_rate)
```
```suggestion
                                                 * std::asinh((current_edot_ii * cellsize ) / (2.0 * quasi_static_strain_rate)
```

anne-glerum replied:
same std:: for all functions here

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+              break;
+            }
+          }
+        // if friction negative, that does not make sense and will get rate and state friction
```
```suggestion
        // A negative friction angle does not make sense and will get rate and state friction
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+          }
+        // if friction negative, that does not make sense and will get rate and state friction
+        // into trouble, so return some very small value
+        if (current_friction < 1e-30)
```
Why not do `std::max(current_friction, 1e-30);`?

anne-glerum reviewed (commented):
Part 1 of friction_options.cc :)

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+      FrictionOptions<dim>::
+      get_theta_composition_mask(ComponentMask composition_mask) const
+      {
+        // Exlude compmositional field "theta" during the volume fraction computation.
```
```suggestion
        // Exclude compositional field "theta" during the volume fraction computation.
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+        // by integration using the assumption that velocities are constant at any time step.
+        double current_theta = critical_slip_distance / ( cellsize * current_edot_ii ) +
+                               (theta_old - critical_slip_distance / ( cellsize * current_edot_ii))
+                               * exp( - ((current_edot_ii * cellsize) * this->get_timestep()) / critical_slip_distance);
```
```suggestion
                               * exp( - (current_edot_ii * cellsize) * this->get_timestep() / critical_slip_distance);
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                               (theta_old - critical_slip_distance / ( cellsize * current_edot_ii))
+                               * exp( - ((current_edot_ii * cellsize) * this->get_timestep()) / critical_slip_distance);
+
+        // TODO: make dt the min theta?
```
Can you add a comment why you apply this cut-off?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                               * exp( - ((current_edot_ii * cellsize) * this->get_timestep()) / critical_slip_distance);
+
+        // TODO: make dt the min theta?
+        if (current_theta < 1e-50)
```
```suggestion
        std::max(current_theta, 1e-50);
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+
+        // TODO: make dt the min theta?
+        if (current_theta < 1e-50)
+          current_theta = 1e-50;
```
```suggestion
          
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+        double cellsize = 1.;
+        if (in.current_cell.state() == IteratorState::valid)
+          {
+            cellsize = in.current_cell->extent_in_direction(0);
```
This line can be moved to the main if (line 292) of this function. Then you can remove this if-statement and line 286.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+      get_theta_composition_mask(ComponentMask composition_mask) const
+      {
+        // Exlude compmositional field "theta" during the volume fraction computation.
+        const unsigned int theta_composition_index = this->introspection().compositional_index_for_name("theta");
```
Why do you have to get `theta_composition_index` here again? I thought you set it once in parse_parameters?

anne-glerum replied:
also add an AssertThrow that get_use_theta() is true

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+            for (unsigned int j=0; j < volume_fractions.size(); ++j)
+              {
+                const double critical_slip_distance = get_critical_slip_distance(in.position[q], j);
+                current_theta += volume_fractions[j] * compute_theta(theta_old, current_edot_ii, cellsize, critical_slip_distance);
```
To me it makes more sense to average the `critical_slip_distance`with the volume_fractions and then compute_theta once afterwards. The result differs slightly. Can't really give arguments.

EstherHeck replied:
I did that now, because I agree that it makes more sense / feels better like this. However, I think initially I put theta into the lop as well, to ave the theta in reaction terms as similar as possible to the theta used to compute the friction angle. In the latter case everything is within such a loop over the volume fractions within calculate_isostrain_viscosities.

anne-glerum replied:
Ah that's also true. In that case keeping it the same as calculate_isostrain_visocities is probably better.

In case you don't change it back, there is no need to declare current_friction before the loop and then add to it; you can do `const double current_friction = compute_theta(...);` after the loop.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+          critical_slip_distance_function->value(Utilities::convert_array_to_point<dim>(point.get_coordinates()),j);
+
+        if (critical_slip_distance <= 0)
+          AssertThrow(false, ExcMessage("Critical slip distance in a rate-and-state material must be > 0."));
```
```suggestion
          AssertThrow(critical_slip_distance > 0, ExcMessage("Critical slip distance in a rate-and-state material must be > 0."));
```
The Asserts are designed to have the condition that should be satisfied inside them.

anne-glerum replied:
Are there similar Asserts you can apply to a and b?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+            | (get_friction_dependence_mechanism() == slip_rate_dependent_rate_and_state_dependent_friction))
+          use_theta = true;
+
+        return use_theta;
```
```suggestion
        const FrictionDependenceMechanism mechanism = get_friction_dependence_mechanism();
         // TODO add the RSF options
        return ((mechanism == rate_and_state_dependent_friction)
                 || (mechanism == regularized_rate_and_state_friction)
                 || (mechanism == slip_rate_dependent_rate_and_state_dependent_friction))
```

anne-glerum replied:
I'dd add all options that use theta & are available now, before they get forgotten.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                                               "regularized rate and state friction"),
+                           "Whether to apply a rate or rate and state dependence of the friction angle. This can "
+                           "be used to obtain stick-slip motion to simulate earthquake-like behaviour, "
+                           "where short periods of high-velocities are seperated by longer periods without "
```
```suggestion
                           "where short periods of high-velocities are separated by longer periods without "
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           "factor', which uses $\\mu* = \\mu\\big(1-\\frac{P_f}{\\sigma_n} \\big)$. "
+                           "Reasonable values for a and b are 0.01 and 0.015, respectively, see \\cite{sobolev_modeling_2017}. "
+                           "In ASPECT the state variable is confined to positive values: if it becomes $<=0$ during the computation "
+                           "it is set to 1e-10. The friction angle is set to 0 if negative. "
```
Line 228 sets the friction angle to 1e-30, not 0. And theta is cut off at 1e-50.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           "where D is the slip at point in fault at the first timestep of earthquake. "
+                           "\n\n"
+                           "\\item ``slip rate dependent rate and state dependent friction'': The rate and state "
+                           "parametera and the critical slip distance L are made slip rate dependent. The friction "
```
```suggestion
                           "parameters and the critical slip distance L are made slip-rate dependent. The friction "
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           "\\item ``slip rate dependent rate and state dependent friction'': The rate and state "
+                           "parametera and the critical slip distance L are made slip rate dependent. The friction "
+                           "coefficient is computed as in 'rate and state dependent friction'. But a and L are not "
+                           "constant, but are computed as follws, see \\citep{Im_im_slip-rate-dependent_2020} for details."
```
```suggestion
                           "constant, but are computed as follows, see \\citep{Im_im_slip-rate-dependent_2020} for details."
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           "\n\n"
+                           "\\item ``regularized rate and state friction'': The friction coefficient is computed using: "
+                           "$\\mu = a\\cdot sinh^{-1}\\left[\\frac{V}{2V_0}exp\\left(\\frac{\\mu_0+b\cdot ln(V_0\theta/L)}{a}\\right)\\right]$ "
+                           "This is a high velocity approxiamtion and regularized version of the classic rate and state friction. "
```
```suggestion
                           "This is a high velocity approximation and regularized version of the classic rate and state friction. "
```
Most IDEs have spelling control as well, does yours?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           "\\cite{pipping_variational_2015} in simple rate-and-state friction models. "
+                           "This parameter only becomes effective when the yield mechanism tresca "
+                           "is specified in the input file. "
+                           "Units: MPa.");
```
Do you convert it to Pa units when parsing? Inside ASPECT, we compute in Pa.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           "is specified in the input file. "
+                           "Units: MPa.");
+
+        /*
```
can this entry be removed?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           Patterns::Double (0),
+                           "The quasi static or reference strain rate used in rate and state friction. It is an "
+                           "arbitrary strain rate at which friction equals the reference friction angle. "
+                           "This happens when slip rate (which is represented in ASPECT as strain rate) "
```
isn't it represented as strain rate * cell size?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+        prm.declare_entry ("Use radiation damping", "false",
+                           Patterns::Bool (),
+                           "Whether to include radiation damping or not. Radiation damping adds the term "
+                           "$-\\etaV = \\frac{\\mu}{2c_s}$ to the yield stress, $\\mu$ is the elastic shear "
```
I'd use a different symbol for mu here, as mu is used to represent the friction coefficient throughout this plugin.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                           "Whether to include radiation damping or not. Radiation damping adds the term "
+                           "$-\\etaV = \\frac{\\mu}{2c_s}$ to the yield stress, $\\mu$ is the elastic shear "
+                           "modulus, $c_s$ the shear wave speed and V the slip rate \\citep{rice_spatio-temporal_1993}. "
+                           "Radiation damping prevents velocities to increase to infinity at the small time steps of "
```
```suggestion
                           "Radiation damping prevents velocities from increasing to infinity at the small time steps of "
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+      void
+      FrictionOptions<dim>::parse_parameters (ParameterHandler &prm)
+      {
+        // Get the number of fields for composition-dependent material properties
```
```suggestion
        // Get the number of fields for composition-dependent material properties
        // including the background field.
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+        if ((friction_dependence_mechanism == slip_rate_dependent_rate_and_state_dependent_friction)
+            |(friction_dependence_mechanism == regularized_rate_and_state_friction)
+            |(friction_dependence_mechanism == rate_and_state_dependent_friction_plus_linear_slip_weakening)
+            |( friction_dependence_mechanism == rate_and_state_dependent_friction))
```
```suggestion
            ||(friction_dependence_mechanism == regularized_rate_and_state_friction)
            ||(friction_dependence_mechanism == rate_and_state_dependent_friction_plus_linear_slip_weakening)
            ||( friction_dependence_mechanism == rate_and_state_dependent_friction))
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                        ExcMessage("The rate and state friction will only work with the nonlinear "
+                                   "solver schemes 'single Advection, single Stokes' and "
+                                   "'single Advection, iterated Stokes', 'single Advection, "
+                                   "iterated defect correction Stokes', 'ginle advection, iterated Newton Stokes'"));
```
```suggestion
                                   "iterated defect correction Stokes', 'single advection, iterated Newton Stokes'"));
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                if (dynamic_angles_of_internal_friction[i] > 90)
+                  {
+                    AssertThrow(false, ExcMessage("Dynamic angles of friction must be <= 90 degrees"));
+                  }
```
```suggestion
                    AssertThrow(dynamic_angles_of_internal_friction[i] <= 90, ExcMessage("Dynamic angles of friction must be <= 90 degrees"));
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+                else
+                  {
+                    dynamic_angles_of_internal_friction[i] *= numbers::PI/180.0;
+                  }
```
```suggestion
                    dynamic_angles_of_internal_friction[i] *= numbers::PI/180.0;
```

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+        // Dynamic friction parameters
+        dynamic_characteristic_strain_rate = prm.get_double("Dynamic characteristic strain rate");
+
+        if (prm.get ("Dynamic angles of internal friction") == "9999")
```
In case the mechanism is not `dynamic_friction` does it matter what this parameter is set to? Maybe add a check for the correct mechanism. On the other hand, do we still need this 9999 check? It's not done for other parameters either; maybe just set the default value to something that makes sense.

EstherHeck replied:
I used this check to set the default value to the internal angle of friction so that it would not matter if not set. But I agree, that a check for the mechanism would be nice

EstherHeck replied:
I could also leave it out completely. I think it was from before I introduced the switch for friction_mechanism? so it was easier to use by setting the default of the dynamic angle to the internal angle. I will take it out then, you are right, that it is not needed anymore.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,849 @@
+            // TODO: make this a bit more flexible name-wise, like let the user define which materials should be
+            // considered. Or which strategy. Could also be all, or take a and b as a proxy.
+            // TODO: should be done if this is > 70 or so %. Can be circumvented right now by using max
+            // composition for viscosity averaging
```
This comment seems out of place here. Can you move it to where it is applicable?

anne-glerum commented on include/aspect/material_model/rheology/visco_plastic.h:
```diff
@@ -0,0 +1,348 @@
+        /**
+         * The value of the rate and state friction parameter a.
+         */
+        std::vector<double> RSF_a;
```
I think the friction_options parameters should go in their own AdditionalOutputs, that are only created when the relevant mechanism is selected.

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+        names.emplace_back("current_friction_angles");
+        names.emplace_back("plastic_yielding");
+
+        // if rate and state friction is used make more additional outputs
```
I'd create a separate make_friction_additional_outputs_names etc

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                    AssertThrow(false, ExcMessage("Usage of radiation damping only makes sense when elasticity is enabled."));
+                  }
+                else
+                  {
```
```suggestion
                    AssertThrow(use_elasticity, ExcMessage("Usage of radiation damping only makes sense when elasticity is enabled."));
```

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                    const double cellsize = current_cell->extent_in_direction(0);
+                    //std::cout << " current edot_ii is " << current_edot_ii << std::endl;
+                    radiation_damping_term = current_edot_ii * cellsize * elastic_shear_moduli[j]
+                                             / (2 * sqrt(elastic_shear_moduli[j] / reference_density));
```
```suggestion
                                             / (2.0 * sqrt(elastic_shear_moduli[j] / reference_density));
```
c++ upgrades integers to doubles when multiplying with a double, but it is very good practice not to rely on this and always write a double when you want a double.

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                                             / (2 * sqrt(elastic_shear_moduli[j] / reference_density));
+                    if (yield_mechanism != tresca)
+                      {
+                        current_stress = current_stress - radiation_damping_term;
```
```suggestion
                        current_stress -= radiation_damping_term;
```

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                    if (yield_mechanism != tresca)
+                      {
+                        current_stress = current_stress - radiation_damping_term;
+                        current_edot_ii = std::max(current_stress / (2 * viscosity_pre_yield), min_strain_rate);
```
```suggestion
                        current_edot_ii = std::max(current_stress / (2.0 * viscosity_pre_yield), min_strain_rate);
```
Note to self: should the radiation term act on the stress invariant (shear stress) or on the yield stress?

EstherHeck replied:
at some point in autumn we talked about that quite often and in the end I applied it to current_stress because current_stress can be used to modify te effective viscosity and current_edot_ii, which in turn modifies the friction angle. Contrary, the yield_stress is only used to be compared to current_stress to determine if we enter yielding, which we always do anyway for RSF. Do you still agree with this reasoning, or do you have another idea?

anne-glerum replied:
Looking at the Erickson ea 2020 paper, eq 1, they include it in the current stress as well, so let's leave it as is.

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                  //std::cout << "current_stress - yield_stress is: " << current_stress - yield_stress  << " in composition " << j<< std::endl << std::endl;
+                  //                   AssertThrow(volume_fractions.size() == in.composition[0].size()+1, ExcMessage("there is a field too much...."));
+
+                  if ((current_stress >= yield_stress) |
```
```suggestion
                  if ((current_stress >= yield_stress) ||
```
You have to be careful with these logical symbols. It might compile, but `|` is only used for bitwise comparison, which is usually not what we do, generally `||` is needed!

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                }
+                case tresca:
+                {
+                  if ((current_stress >= yield_stress) |
```
```suggestion
                  if ((current_stress >= yield_stress) ||
```

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                      // In \\cite{pipping_variational_2015} it is stated that this is the
+                      // equation for Tresca friction
+                      double fault_strength = friction_options.effective_normal_stress_on_fault
+                                              * tan(output_parameters.current_friction_angles[j]) * current_edot_ii
```
```suggestion
                                              * std::tan(output_parameters.current_friction_angles[j]) * current_edot_ii
```

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                      // the fault strength is equal to the shear stress on the fault.
+                      // In \\cite{pipping_variational_2015} it is stated that this is the
+                      // equation for Tresca friction
+                      double fault_strength = friction_options.effective_normal_stress_on_fault
```
```suggestion
                      const double fault_strength = friction_options.effective_normal_stress_on_fault
```

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                    }
+                  break;
+                }
+                case tresca:
```
I don't think this tresca implementation is correct. The yield stress is calculated above with the drucker-prager criterion, but this should be the fault_strength used here. We switch out the DP criterion for the one from the benchmark, not use both of them somehow.

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+                                              * tan(output_parameters.current_friction_angles[j]) * current_edot_ii
+                                              * current_cell->extent_in_direction(0)-radiation_damping_term;
+
+                      current_edot_ii = fault_strength / (2 * viscosity_pre_yield);
```
```suggestion
                      
```

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+
+        // If friction is defined state dependent, the material field for the state variable theta
+        // must be excluded during volume fraction computation.
+        if (friction_options.get_use_theta())
+          composition_mask.set(friction_options.theta_composition_index,false);
```
friction_options has a function `get_theta_composition_mask`, please use that.

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -0,0 +1,893 @@
+        constant_viscosity_prefactors.parse_parameters(prm);
+
+        // Plasticity parameters
+        drucker_prager_parameters = drucker_prager_plasticity.parse_parameters(this->n_compositional_fields()+1,
```
```suggestion
        drucker_prager_parameters = drucker_prager_plasticity.parse_parameters(n_fields,
```

anne-glerum reviewed (commented):
Hi Esther, there are many comments related to spelling, clarification of comments or 'syntactic sugar', but several things really need to be changed. Here's what I can remember (I've indicated it all of course):
1. the effective_friction_factor of the dynamic friction
2. change the bitwise operator `|` to the logical operator `||`
3. use the benchmark `fault_strength` instead of the drucker-prager yield criterion for the new tresca option

EstherHeck committed:
WIP to get all params needed

EstherHeck committed:
addressed all of Annes comments about friction_options.h, renamed
regularized RSF

EstherHeck committed:
addressed some comits from Anne

EstherHeck committed:
adressed all easy comments from Anne

EstherHeck committed:
introduced parameter use_always_yielding

EstherHeck committed:
rearranged compute_dependent_friction to avoid code duplication

EstherHeck committed:
documentation and removable of outdated check for dynamic rfiction angles

EstherHeck committed:
assert for correct tan/atan conversion

EstherHeck committed:
FrictionAdditionalOutputs working and  documentation things

EstherHeck committed:
a function to compute delta_theta_max

EstherHeck committed:
Lapusta time step builds

EstherHeck committed:
segmentation fault error when trying to get the component mask in the lapusta time step

EstherHeck committed:
all four Lapusta timestep criteria added. Still some todos and runtime error

EstherHeck committed:
compiling but not running, moved delta_theta_max function to visco_plastic

EstherHeck committed:
last changes after Annes review, assertthrow for friction nan or inf

EstherHeck committed:
assert for inf or nan delta_theta_max

EstherHeck committed:
something in the check prints

EstherHeck committed:
return something after switch to get rid of compile warning

EstherHeck committed:
added steady-state friction as a friction option, but not sure if I take the right velocity

EstherHeck committed:
still not working. Cannot call rheology function from material model.

EstherHeck committed:
added parameter whether to cut edot_ii after RDT or not

EstherHeck committed:
for a<=0 still take into account contrib from b

EstherHeck committed:
Lapusta time step does not through error anymore

EstherHeck committed:
if min healing dt <= 0 make it very large

EstherHeck committed:
dont let volume_fractions become negative

EstherHeck committed:
use constant not actual velocity in stst friction

EstherHeck committed:
remove wrong )

EstherHeck committed:
uncomment min healing time step to see if Lapusta only does not work bc of theta

EstherHeck committed:
made it work again after rebase

EstherHeck committed:
limit theta_old to >=0)

EstherHeck committed:
limit old theta to >= 1e-50

EstherHeck committed:
steady-state RSF now also within use_theta, though not sure if it should be

EstherHeck committed:
added effective friction factor in steady state RSF and made indent

EstherHeck committed:
reintroduction of theta_increment to limit it so the new theta in reaction terms remains positive

EstherHeck committed:
made single parameter point out of point_a and point_b

EstherHeck committed:
documentation and improvement for steady-state RSF, theta now changes appropriately,

made effective friction factor a function input

EstherHeck committed:
prototype particle property theta builds

EstherHeck committed:
make friction assert more informative

EstherHeck committed:
not cut old theta, particle property theta working

cut theta on particle at 1e-50 and cut theta increment

compute theta with min SR if not yielding

clean up comments

EstherHeck committed:
all lapusta timesteps taken into account, some cleaning up

EstherHeck committed:
assertthrow for negative theta in min healing timestep function

EstherHeck committed:
reduced effective friction factor to one function and made it pf/sig_n without the 1-

EstherHeck committed:
Lapusta timestep length written to screen

EstherHeck committed:
added effective friction factor in regularized RSF for a=0, made indent

EstherHeck committed:
add prints, no cut old theta, do cut current theta and increment

EstherHeck committed:
print neg theta on particle

EstherHeck committed:
do cut old theta and increment, not cut current theta

EstherHeck committed:
not cut theta increment

EstherHeck committed:
do cut old, current, and increment theta

EstherHeck committed:
compute RDT when using tresca yield option

EstherHeck committed:
outcomment asserts that inhibit iterated advection

EstherHeck committed:
if min helaing dt negative, dont use it

EstherHeck committed:
adjusted error message for effective friction factor >=0 and <1

EstherHeck committed:
return largest number for healing dt if it is <= 1e-50

EstherHeck committed:
also print if theta is positive

EstherHeck committed:
esle to else

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,997 @@
+                  // theta_old is taken from the current compositional field theta
+                  const double theta_old = composition[theta_composition_index];
+
+                  // if we do not assume always yielding, then theta should not be updated
```
Why? I don't understand this.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,997 @@
+                  const double effective_friction_factor = get_effective_friction_factor(position,j);
+
+                  // theta_old is taken from the current compositional field theta
+                  const double theta_old = composition[theta_composition_index];
```
cut theta_old

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,997 @@
+                                                          average_elastic_shear_moduli, use_elasticity,
+                                                          use_reference_strainrate, dte);
+
+            const double theta_old = in.composition[q][theta_composition_index];
```
cut theta_old at zero

EstherHeck committed:
print the position where theta is negative

EstherHeck committed:
made theta = old_theta for very low strain rates, also print zero theta, made indent

EstherHeck committed:
only compute theta when inside fault material, and only update theta on particle if not in dt 0

EstherHeck committed:
make prints more informative: Zero and dt

EstherHeck committed:
only compute theta when in fault material of >50% volume, same for going into compute_dependent_friction_angle, only compute theta once, in reaction terms, some other things that came along

EstherHeck committed:
make an input parameter whether to print negative/Zero theta

EstherHeck committed:
make it work after rebase

EstherHeck committed:
make it compile after rebase with master, make indent

EstherHeck committed:
added an RSF convection time step. Made indent

anne-glerum commented on source/time_stepping/lapusta_time_step.cc:
```diff
@@ -0,0 +1,183 @@
+                // to capture the increasing slip rate in case of a purely rate-dependent friction, i.e. if b = 0
+                min_vep_relaxation_time_step = std::min (min_vep_relaxation_time_step,
+                                                         0.2 * out.viscosities[q]
+                                                         / viscoplastic.get_elastic_shear_modulus(in.composition[0]));
```
why composition[0]? Think it should be q again.

EstherHeck replied:
thanks for spotting this! I think the 0 is there because I tried to find out in other plugins how to access values and somewhere I saw and copied the 0

anne-glerum commented on source/time_stepping/lapusta_time_step.cc:
```diff
@@ -0,0 +1,183 @@
+                const double local_velocity = velocity_values[q].norm();
+
+                std::pair<double,double> delta_theta_max_and_critical_slip_distance = viscoplastic.compute_delta_theta_max(
+                                                                                        in.composition[0],  // should this be 0 or q? for all the times I hand over composition.. found 0 in some part of the code, but q intuitively makes more sense
```
should be q as that's the current quad point. Otherwise you're using the 0th quad point for all the quad points in the cell.

anne-glerum commented on source/time_stepping/lapusta_time_step.cc:
```diff
@@ -0,0 +1,183 @@
+            // with Delta d_max = 1e-3 in Herrendörfer et al. 2018
+            // here, the term  min(|Delta x/v_x|,|Delta x/v_y|) is simplified to min Delta x / max_local_velocity
+            min_displacement_time_step = std::min (min_displacement_time_step,
+                                                   1.e-3 * cell->minimum_vertex_distance() / max_local_velocity);
```
above `delta_x` was defined as `minimum_vertex_distance` already

anne-glerum commented on source/time_stepping/lapusta_time_step.cc:
```diff
@@ -0,0 +1,183 @@
+
+      // take the minimum of the four criteria
+      // TODO: is there a more elegant way to take the minimum of four values?
+      min_lapusta_timestep = std::min (min_state_weakening_time_step,
```
Instead of having 4 doubles, you could have a vector `timestep` of 4 doubles. Then you could do `min_lapusta_timestep = *std::min_element(timestep.begin(), timestep.end());`

EstherHeck replied:
thanks!

anne-glerum commented on source/time_stepping/lapusta_time_step.cc:
```diff
@@ -0,0 +1,183 @@
+                                       std::min (min_healing_time_step,
+                                                 std::min (min_displacement_time_step,
+                                                           min_vep_relaxation_time_step)));
+
```
It seems you don't communicate the minimum timestep between the processes. You need something like:
```
const double min_global_lapusta_timestep
 = Utilities::MPI::min (min_lapusta_timestep, this->get_mpi_communicator());
```

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -307,6 +371,100 @@ namespace aspect
+    ViscoPlastic<dim>::
+    compute_min_healing_time_step (const std::vector<double> &composition) const
+    {
+      double min_healing_time_step = 0.2 * composition[rheology->friction_options.theta_composition_index];
```
Cut off composition at 0

EstherHeck committed:
adressed Annes quick Lapusta time step review. Deleted two folders in doc/manual/cookbook/benchmarks to see if that gets rid of that rebase error message with the hints of editing the todo

EstherHeck committed:
adressed Annes quick Lapusta time step review

EstherHeck committed:
time in point values set to higher precision when RSF is used. Made indent

EstherHeck committed:
only take higher CFL number if velocity is significantly (100) lower

EstherHeck committed:
only take higher CFL number if velocity is higher (5 times)

anne-glerum commented on source/time_stepping/RSF_convection_time_step.cc:
```diff
@@ -0,0 +1,143 @@
+
+      // if max velocities stayed the same or decreased, take the CFL number specified in the input file.
+      // else, take 0.5 as the CFL number to let the timestep increase faster
+      if ((max_global_speed_over_meshsize != 0.0) && (max_global_speed_over_meshsize <= max_global_speed_over_meshsize_old))
```
I don't understand this logic. If the current velocity is smaller than the previous velocity, isn't that the moment when you want the timestep to increase faster?

EstherHeck replied:
You are completely right, thanks for spotting this! I wrote it down for the timestep length, which becomes smaller when we want the smaller CFL number. And somehow I wrongly made it into a velocity rule...

anne-glerum commented on source/time_stepping/RSF_convection_time_step.cc:
```diff
@@ -0,0 +1,143 @@
+                                        "and temperature polynomial degree. Depending on the current "
+                                        "and previous velocity, different CFL numbers are used to "
+                                        "increase timestep size faster after an earthquake. This is "
+                                        "only recommended for rate-and-state friction models.")
```
In this case, you should throw when RSF is not used.

EstherHeck replied:
good point, thanks!

EstherHeck committed:
take the right logic for when to increase CFL number

EstherHeck committed:
lapusta tiestep only computed within fault. Made indent

EstherHeck committed:
introduced only temporarily important parameter to use min time step length for the first timestep

EstherHeck committed:
only compute theta on particle if initial fault value > 0.5

EstherHeck committed:
fixed double bool mistake

EstherHeck committed:
fix exception for theta on particle - fault not initial fault

EstherHeck committed:
try the other way to ask for fault index on particle

EstherHeck committed:
Part 1: update comments and todos, go through logic

EstherHeck committed:
Part 2: update comments and todos, go through logic

EstherHeck committed:
Part 3 (last part): update comments and todos, go through logic

EstherHeck committed:
refined theta prints

EstherHeck committed:
also print particle ID for checking the theta values

EstherHeck committed:
more and reworked prints for theta, made indent

EstherHeck committed:
lots of changes in the theta printing and ONE IMPORTANT change in the particle theta updating, theta reaction terms NOT using theta equation for testing purpose

EstherHeck committed:
compute_theta_reaction terms now updates theta correctly again

EstherHeck committed:
asserts for current/old theta <= 0 wherever used

EstherHeck committed:
renamed "quasi static strain rate" parameter to "RSF reference slip rate"

EstherHeck committed:
todo for default for critical slip distance function

EstherHeck committed:
corrected error message for old theta

EstherHeck committed:
overwrite stresses on particles with particle value

EstherHeck committed:
provisoricly moved effective friction factor from friction to pressure

EstherHeck committed:
added effective friction factor to lapusta timesteps delta_theta_max, removed (1- from effective friction factor at yielding,ran made indent

EstherHeck committed:
dont print stress values to screen

EstherHeck committed:
ran make indent

EstherHeck committed:
added "1-" to the get_effective_friction_factor function to ensure that default of 0 makes sense

EstherHeck committed:
also put the "1-" into the assertthrow...

EstherHeck committed:
make 1 to 0 - undo a mistake and crudely make lapusta hopefully work

EstherHeck committed:
print infos for min vep relaxation time step

EstherHeck committed:
removal of the prints in the lapusta time step and add comments

EstherHeck committed:
moved RDT into drucker prager yield stress

EstherHeck committed:
fixed assert

EstherHeck committed:
removed outdated parameter cut_edot_ii after RDT

EstherHeck committed:
print yield stress current stress viscosity to check if it converges

EstherHeck committed:
outcomment printing for vep relaxation time step

EstherHeck committed:
add prints for different yield stresses

EstherHeck committed:
removed unecessary parameter print_thetas, made indent

EstherHeck committed:
removed unused function

EstherHeck committed:
modified stress prints, removed unused variable

EstherHeck committed:
refine print

EstherHeck committed:
do something with prints

EstherHeck committed:
hopefully fix the issue that critical slip distance function must be specified by moving assert into if(use_theta)\n to avoid complications, also only use L from fault material to compute theta

EstherHeck committed:
restructuring of lapusta timestep

EstherHeck committed:
Fixed: RSF parameter function now only need to be specified if RSF is used

EstherHeck committed:
remove deprecated parameter effective_friction_factor and clean up documentation for effective_friction_factor_function

EstherHeck committed:
removed deprecated todos

EstherHeck committed:
fixed critical slip distance assert

EstherHeck committed:
go through parameter texts

EstherHeck committed:
made indent

EstherHeck committed:
somment on units for theta

EstherHeck committed:
deleted all cutting of theta except for within compute_theta and on particle. Cleaned up comments in particle property theta

anne-glerum committed:
Update VEP implementation

anne-glerum committed:
Add TODOs

anne-glerum committed:
Update TODOs based on discussion and reviews

EstherHeck committed:
make an input list for what compositions should use RSF

anne-glerum committed:
Update TODOs

anne-glerum committed:
Add reaction_rates

anne-glerum committed:
Update according to App E

anne-glerum committed:
Add ReactionRates

anne-glerum committed:
Update fill_reaction_terms

anne-glerum committed:
Update parse_parameters

anne-glerum committed:
Update calculate_isostrain_viscosities

anne-glerum committed:
Assert no plastic damper

anne-glerum committed:
Update comments

anne-glerum committed:
Fix 2 bugs

anne-glerum committed:
Add pp to output the min and max of the stress components

EstherHeck committed:
fix a debug-runtime error about the RSF-mask

EstherHeck committed:
work on friction/theta asserts

anne-glerum committed:
Fix precision statistiscs

EstherHeck committed:
rename variables after rebase etc

EstherHeck committed:
get old theta correctly - using elastic stresses as an example

EstherHeck committed:
WIP get old theta correctly, building with errors

EstherHeck committed:
get old theta correctly - no build error left

EstherHeck committed:
tested that old_theta correct and fixed runtime errors after merge

EstherHeck committed:
only compute RSF if in RSF material and small other things

anne-glerum commented on source/material_model/rheology/friction_models.cc:
```diff
@@ -359,6 +364,7 @@ namespace aspect
             && in.current_cell.state() == IteratorState::valid)
           {
             const double fault_volume = get_fault_volume(volume_fractions);
+            // ToDo: make an input parameter, how much fault there needs to be?
```
Actually, I'm wondering whether instead of checking whether the fault_volume is larger than 0.5, you should just only fill the reaction term for the RSF fields that could have a nonzero theta (`RSF_composition_masks`), like you do below. So remove the fault_volume>0.5 completely. Would that give problems outside of the fault because theta could/would be 0 there? Then there could be a check that if theta is zero, then the reaction term is 0, or something.

EstherHeck replied:
hm, I think I cant really follow your thoughts here. 
The `fault_volume `is based on the `RSF_composition_masks` by adding up the volume fractions of all RSF_compositions. So I do only fill the reaction terms for theta where it could be non-zero.

anne-glerum replied:
Ah I see, but that means you don't fill the reaction terms for where theta is non-zero, but for where the fault volume is larger than 0.5. They could be slightly different things right? Because there is still some fault material even if 0>fault_volume<0.5. Therefore I was wondering if you can compute the reaction terms for all the RSF_composition_masks, regardless of whether their combined fault_volume>0.5 or not. Outside the fault, updates to theta should be small because of the strain rate/velocity being small right?

EstherHeck replied:
* theta-updates-non-zero. Theta should never be zero, because then friction is nan

Yes, they could be slightly different things. I was considering to make the 0.5 an input paramter, so I could variably include more and less faulty areas. 
The thing is, if I simply include all the model, then computation became very slow (at least a good year ago). 
Furthermore, theta is based on slip rate and not strain rate in the actual RSF philosophy. So outside the fault, the velocities/strain rates in aspect are high, so, the way it is implemented theta would change a lot. But they are not slip rates on a fault surface, so in theory theta should not change much there as it displays a bit the time since the last movement on a fault surface. I could maybe change the computation and say that theta should be computed wherever plastic_yielding occurs to account for this. What do you think? 
But then we are again at the problem, that the entire fault should always be yielding according to RSF theory

EstherHeck committed:
work on Annes comments

EstherHeck committed:
cut theta instead of assert

EstherHeck committed:
resolve segmentation fault error with checking that not in bg field

EstherHeck committed:
uncomment stress_averaging and remove stress_averaging time_scale_factor from assert

EstherHeck committed:
uncomment atan assert

anne-glerum committed:
Fix merge leftovers

anne-glerum committed:
Output full stress in visu pp

anne-glerum committed:
Add includes

anne-glerum committed:
Rm include

anne-glerum committed:
Fix shear stress pp

anne-glerum committed:
Make sure particle update occurs at start timestep

anne-glerum committed:
Fix Assert

anne-glerum committed:
Do not during particle update before t1

anne-glerum committed:
Add Assert to visu plugins

anne-glerum committed:
Check that composition particle property is not used

anne-glerum committed:
Only compute stress tensor residuals when elasticity is on

anne-glerum committed:
Fix after merge

anne-glerum committed:
Address two review comments

anne-glerum committed:
Check that ve_stress_xx exists before asking field nr

anne-glerum committed:
Rm removed pp from prm files

anne-glerum committed:
Fix tests

gassmoeller committed:
Merge branch 'main' into fix_stresses_elasticity

EstherHeck committed:
cleaning up after rebase onto PR4370 on 29th nov 2023 commit 30df10c. e g renaming variables"

EstherHeck committed:
made indent and resolved first build errors

EstherHeck committed:
found two typos
